### DISCUSSION TITLE
avrdude 7.2 binaries for Arduino platforms

### DISCUSSION THREAD

**Discussion by SpenceKonde** [avrdudes/avrdude#1477] [Q&A] [answered]
1. point users to in my boardmanager.json file for the 7 holy platforms? I see a few platforms in  releases, but I need all the 7 Arduino platforms, otherwise people get angry at me.  When Arduino stopped building avrdude for x86 linux I had to workaround it by making a release of avrdude that pointed other platformss to the new version and x86 linux to the old version. (People go from gentle and harmless to biting, scratching and furious balls of teeth, spikes, and claws if they feel like they're being cornered, and that's what you do  when you threaten to drop support for their OS. They froth at the mouth, When the matter was discussed in person (thankfuklly I was merely a witness) such a discussion nearly came to blows. What I'm saying is, I'm not going to drop an OS version, and in the thread that has vanished, you folks indicated that all of them would exist.... 

3. Or at least someplace that I can download it from and rehost if you don't want to deal with the traffic costs;. (like my toolchain policy - go ahead and use my azduino toolchains - but don't just go pointing at my URLs unless they're free ones (I only "freehost" the linux x64 avr-gcc in github, the rest are in S3 and I have to pay some tiny fee per download of it. The linux one is freehosted because >95% of the toolchain downloads in total are downloaded by github. Keeping that traffic in network saves them money and it sure saves me money. 

The other cores I see only seem to be using 7.1, and I specifically wanted a few of the things in 7,2. 

Thanks

At this point I'm ready to switch my cores over to it

**Comment by stefanrueger** [comment]
> [files for the] 7 Arduino platforms

Do you mean AVRDUDE binaries for operating systems that support an Arduino IDE? Which ones do you miss? BSD, FreeBSD, NetBSD, OpenBSD, DragonFly, BSD, NeXTSTEP, Darwin (macOS), iOS, tvOS, watchOS, bridgeOS, SunOS, Ultrix, Android, Arch, ChromeOS, Debian, Fedora, Gentoo, Red Hat, SUSE, Ubuntu, System V, A/UX, AIX, HP-UX, IRIX, OpenServer, Solaris, OpenSolaris, Illumos, Tru64 UNIX, UnixWare?

Or do you mean hardware platforms? Intel, PowerPC, ARM, ...

**Comment by MCUdude** [comment]
I think these binaries from @mcuee are statically built and suitable for bundling.
https://github.com/mcuee/avrdude/releases/tag/v7.2

And I think Arduino is ready to release their 7.2 version soon:
https://github.com/arduino/avrdude-packing/issues/28

> At this point I'm ready to switch my cores over to it

 Even replacing pymcuprog and the micronucleus uploader? Avrdude's serialupdi implementation has turned out to be flawless, with no serious bugs reported since  v7.0

**Reply by mcuee** [reply]
> I think these binaries from @mcuee are statically built and suitable for bundling.
> https://github.com/mcuee/avrdude/releases/tag/v7.2

The following unofficial binaries are not that suitable for bundling -- as I am not really going to support them...
They are more suitable for testing. In fact, they are actually using one commit after the official v7.2 release (same as git main https://github.com/avrdudes/avrdude/commit/330030ed234bb2318fcc0e7d3f1805d1ab89345e) as I am lazy to change the github action script.

[avrdude_v7.2_Linux_32bit.tar.gz](https://github.com/mcuee/avrdude/releases/download/v7.2/avrdude_v7.2_Linux_32bit.tar.gz)

[avrdude_v7.2_Linux_64bit.tar.gz](https://github.com/mcuee/avrdude/releases/download/v7.2/avrdude_v7.2_Linux_64bit.tar.gz)

[avrdude_v7.2_Linux_ARM64.tar.gz](https://github.com/mcuee/avrdude/releases/download/v7.2/avrdude_v7.2_Linux_ARM64.tar.gz)

[avrdude_v7.2_Linux_ARMv6.tar.gz](https://github.com/mcuee/avrdude/releases/download/v7.2/avrdude_v7.2_Linux_ARMv6.tar.gz)

[avrdude_v7.2_macOS_64bit.tar.gz](https://github.com/mcuee/avrdude/releases/download/v7.2/avrdude_v7.2_macOS_64bit.tar.gz)

[avrdude_v7.2_Windows_32bit.tar.gz](https://github.com/mcuee/avrdude/releases/download/v7.2/avrdude_v7.2_Windows_32bit.tar.gz)

**Comment by mcuee** [comment]
@SpenceKonde 

I will suggest you use official Windows x86/x64 binraries from avrdude project, and then wait for Arduino-packing project for Linux (x86/x64/ARM64/ARMv6) and macOS x64 binaries.
https://github.com/arduino/avrdude-packing/issues/28

**Reply by SpenceKonde** [reply]
Oh, so 7.2 is released only for individual users on certain platforms and is not yet available for distribution with a core? Okay then. I will continue using AVRdude 6.3 for the next release of  DxC and mTC

**Reply by stefanrueger** [reply]
To me it makes sense that those platforms that want AVRDUDE package it in the way they need it. @mcuee's tarballs demonstrate that the project compiles nicely.

**Reply by mcuee** [reply]
The avrdude-packing project shows that it can be done -- using containers to build the static linked binaries for Linux and macOS. I am just using the same container to build for git main in my fork of the avrdude-packing.

**Reply by mcuee** [reply]
On the other hand, the patches adopted by the avrdude-packing project is not really acceptable as of now.
https://github.com/arduino/avrdude-packing/tree/main/patches
006 patch is probably okay (may need the review from @mariusgreuel)
007 patch is actually just a hack and not really correct
008 patch is Arduino branding which is not necessary.

**Reply by mcuee** [reply]
From avrdude project point of view, there is still an issue to be sorted out before we can produce Linux and macOS static linking binaries.
* https://github.com/avrdudes/avrdude/issues/1159

After that we can probably using a container (similar to the one used by Arduino-packing project) to build official binaries for Linux and macOS, if there is a concensus from the admins to support this. This is a niche use case as the Linux distros and macOS Homebrew/macports project will provide the binaries for most users. But I understand that Arduino compatible core developers (or some IDE developers) would like to get such binaries. Personally I think it is okay to leave that to the Arduino-packing project. But I am fine with the othe way if the admins would like to support this use case.
* https://github.com/avrdudes/avrdude/discussions/1433

**Reply by MCUdude** [reply]
> I will continue using AVRdude 6.3 for the next release of DxC and mTC

Why not just bundle [Avrdude 7.1-arduino.1](https://github.com/arduino/avrdude-packing/releases/tag/7.1-arduino.1) while waiting for 7.2? That's what I've been doing for the last six months. 7.1 is still much, much better than 6.3.

Just curious, what is it that 7.2 has that's important, that 7.1 doesn't have?

**Reply by mcuee** [reply]
I agree with @MCUdude that it is better to bundle 7.1 version first.

**Reply by dl8dtl** [reply]
AVRDUDE is open source: it releases sourcecode ¹). It is up to the maintainers of OS packages or other 3rd-party distributors (like Arduino) to compile it, and provide binary releases fitting into their respective environment.

¹) Actually, since with the discontinuation of Eric Weddington's "WinAVR", it turned out nobody seemed willing to prepare Windows binaries for a long time. Thus, I ended up in cross-compiling them on my FreeBSD, and offering those for download. Nowadays, with Github's infrastructure, it is also possible for us to easily provide pre-compiled binaries for certain OS environments which we use for CI testing anyway. But they are not supposed to be a substitute for OS integrators to rather deliver something that fits into their respective packaging strategy.

**Comment by mcuee** [comment]
Updated on 25-Oct-2023

Here is the updated situation for avrdude version used by major OS distributions. We will rely on the community to raise ticket to the respective project/distro to upgrade the avrdude version in the distributions.

1) macOS Homebrew (avrdude 7.2)
https://github.com/Homebrew/homebrew-core/blob/0328ea99163b8fd854468908f76ad2a22f126a74/Formula/avrdude.rb

2) Arch Linux has avrdude 7.2
https://archlinux.org/packages/extra/x86_64/avrdude/

3) FreeBSD ports (avrdude 7.2)
https://www.freshports.org/devel/avrdude

4) macOS macPorts (avrdude 7.0)
https://ports.macports.org/port/avrdude/details/

5) OpenBSD ports (avrdude 6.3)
https://openports.pl/path/devel/avrdude

6) NetBSD pkgsrc (avrdude 6.4)
https://ftp.netbsd.org/pub/pkgsrc/current/pkgsrc/cross/avrdude/index.html

7) WIndows winget -- avrdude 7.1
https://github.com/microsoft/winget-pkgs/tree/master/manifests/a/AVRDudes/AVRDUDE/7.1

8) Windows Scoop -- avrdude 7.2
https://github.com/ScoopInstaller/Main/blob/master/bucket/avrdude.json

9) Windows chocolatey -- avrdude 6.3
https://community.chocolatey.org/packages/avrdude

10) Debian Linux stable and testing -- avrdude 7.1
https://packages.debian.org/search?keywords=avrdude

11) Ubuntu 23.04 and earlier use avrdude 6.3, Ubuntu 23.10 uses avrdude 7.1
https://packages.ubuntu.com/search?keywords=avrdude

12) Fedora 37/38/39/40 are using avrdude 6.4
https://src.fedoraproject.org/rpms/avrdude

13) Haiku ports -- avrdude 7.0 
https://depot.haiku-os.org/#!/pkg/avrdude/haikuports/haikuports_x86_64/7/0/-/-/2/x86_64?bcguid=bc261-HDNA

14) DragonFly BSD -- avrdude 7.1
https://github.com/DragonFlyBSD/DPorts/tree/master/devel/avrdude

15) Gentoo -- avrdude 7.0 
https://packages.gentoo.org/packages/dev-embedded/avrdude

16) Opensuse Tumbleweed -- avrdude 7.2
https://software.opensuse.org/package/avrdude

17) Other distros
https://pkgs.org/download/avrdude

**Comment by mcuee** [comment, marked as answer]
Arduino avrdude-packing project has released their version of Arduino 7.2, which should be suitable for the usage here.
https://github.com/arduino/avrdude-packing/releases/tag/7.2-arduino.1
